File: Solution/Javascript/72. Edit Distance.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
var minDistance = (word1, word2) => {
    let lenA = word1.length;
    let lenB = word2.length;
    let d = createMatrix(lenB + 1, lenA + 1);

    for (let i = 0; i <= lenA; i++) {
        d[i][0] = i;
    }

    for (let i = 0; i <= lenB; i++) {
        d[0][i] = i;
    }

    for (let i = 1; i <= lenA; i++) {
        for (let j = 1; j <= lenB; j++) {
            if (word1[i - 1] == word2[j - 1]) {
                d[i][j] = d[i - 1][j - 1];
            } else {
                d[i][j] = Math.min(d[i - 1][j] + 1, d[i][j - 1] + 1, d[i - 1][j - 1] + 1);
            }
        }
    }

    return d[lenA][lenB];
}

let createMatrix = (rowNum, colNum) => {
    let matrix = [];
    matrix.length = colNum;
    for (let i = 0; i < matrix.length; i++) {
        matrix[i] = [];
        matrix[i].length = rowNum;
        matrix[i].fill(0);
    }
    return matrix
}

let a = "abcde";
let b = "abcd";
console.log(editDistance(a, b));

File: Solution/Python/72. Edit Distance.py
class Solution:
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        lenA = len(word1)
        lenB = len(word2)
        d = [[0 for i in range(lenB + 1)] for i in range(lenA + 1)]

        for i in range(len(d)):
            d[i][0] = i
        for i in range(len(d[0])):
            d[0][i] = i

        for i in range(1, lenA + 1):
            for j in range(1, lenB + 1):
                if (word1[i - 1] == word2[j - 1]):
                    d[i][j] = d[i - 1][j - 1]
                else:
                    d[i][j] = min(d[i - 1][j] + 1, d[i][j - 1] + 1, d[i - 1][j - 1] + 1)

        return d[lenA][lenB]


s = Solution()
word1 = "abcde"
word2 = "abcd"
res = s.minDistance(word1, word2)
print(res)
